File: string/zigzag_conversion.cpp
#include<iostream>
#include<vector>
#include<string>

using namespace std;

class Solution {
public:
    string convert(string s, int numRows) {
        if (numRows == 1 || numRows >= s.length()) return s;

        vector<string> res (numRows, "");
        int rowIndex = 0;
        int step = 1;

        for (char const &letter : s) {
            if (rowIndex == 0) 
                step = 1;
            if (rowIndex == numRows - 1) 
                step = -1;

            res[rowIndex] += letter;
            rowIndex += step;
        }

        string output;
        for (auto &str : res) 
            output += str;
        
        return output;
    }
};

int main() {
    Solution sol = Solution();
    cout << sol.convert("PAYPALISHIRING", 3) << endl;
}

File: string/zigzag_conversion.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1 or numRows >= len(s):
            return s
        res = [""] * numRows
        rowIdx = 0
        step = 1

        for letter in s:
            # if we reach index 0, we have to start moving down the rows
            if rowIdx == 0:
                step = 1
            # if we reach the last row, we have to start moving up the rows
            if rowIdx == numRows - 1:
                step = -1

            res[rowIdx] += letter
            rowIdx += step

        return "".join(res)
